lib/tests/streamlit/session_state_test.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2018-2021 Streamlit Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+from unittest.mock import MagicMock, patch
+
+from streamlit.errors import StreamlitAPIException
+from streamlit.report_thread import _StringSet
+from streamlit.session_state import SessionState
+
+
+# TODO: Mock widget values and modify the tests below to account for this once
+#       session_state and widget state are unified.
+class SessionStateTests(unittest.TestCase):
+    def setUp(self):
+        self.session_state = SessionState(
+            old_state={"foo": "bar"},
+            new_state={"baz": "qux"},
+        )
+
```

Comment:
We should also test the invariants defined in https://www.python.org/dev/peps/pep-3119/#mappings, just to make sure it behaves as expected

Reply:
Done! I added a small test that essentially takes the mapping invariants written out in pep-3119 and rewrites them in a slightly more concise way.

This feels like a place where it'd be nice/appropriate to write some property-based tests, but that might be something we want to do when less time constrained

Reply:
Yeah, I had the same thought. Adding tests for invariants, even if we only test them for a few explicit examples, is still good, and coming up with properties is the hard part of property tests IME. Writing some generators and plugging them in can easily happen whenever we have a free few hours.